File: dart/planet-time/bin/fixture_runner.dart
// fixture_runner.dart — standalone executable
// Reads reference.json and validates 54 cross-language fixture entries.
//
// Usage:
//   dart run bin/fixture_runner.dart ../../c/planet-time/fixtures/reference.json

import 'dart:convert';
import 'dart:io';
import 'package:interplanet_time/interplanet_time.dart';

void main(List<String> args) {
  final path = args.isNotEmpty
      ? args[0]
      : '../../c/planet-time/fixtures/reference.json';

  final file = File(path);
  if (!file.existsSync()) {
    print('SKIP: fixture file not found at $path');
    print('0 passed  0 failed  (fixtures skipped)');
    exit(0);
  }

  final Map<String, dynamic> json;
  try {
    json = jsonDecode(file.readAsStringSync()) as Map<String, dynamic>;
  } catch (e) {
    stderr.writeln('Failed to parse fixture: $e');
    exit(1);
  }

  final entries = (json['entries'] as List<dynamic>).cast<Map<String, dynamic>>();

  int passed = 0;
  int failed = 0;

  for (final entry in entries) {
    final int utcMs = (entry['utc_ms'] as num).toInt();
    final String planetStr = entry['planet'] as String;
    final int expHour = (entry['hour'] as num).toInt();
    final int expMinute = (entry['minute'] as num).toInt();
    final double? expLightS = entry['light_travel_s'] != null
        ? (entry['light_travel_s'] as num).toDouble()
        : null;
    final int expPiw = entry['period_in_week'] != null
        ? (entry['period_in_week'] as num).toInt()
        : -1;
    final int expIsWorkPeriod = entry['is_work_period'] != null
        ? (entry['is_work_period'] as num).toInt()
        : -1;
    final int expIsWorkHour = entry['is_work_hour'] != null
        ? (entry['is_work_hour'] as num).toInt()
        : -1;

    final tag = '$planetStr@$utcMs';

    Planet planet;
    try {
      planet = Planet.fromString(planetStr);
    } catch (e) {
      failed++;
      print('FAIL: $tag unknown planet: $planetStr');
      continue;
    }

    final pt = getPlanetTime(planet, utcMs);

    if (pt.hour == expHour) {
      passed++;
    } else {
      failed++;
      print('FAIL: $tag hour=$expHour (got ${pt.hour})');
    }

    if (pt.minute == expMinute) {
      passed++;
    } else {
      failed++;
      print('FAIL: $tag minute=$expMinute (got ${pt.minute})');
    }

    if (expLightS != null && expLightS != 0.0 &&
        planetStr != 'earth' && planetStr != 'moon') {
      final lt = lightTravelSeconds(Planet.earth, planet, utcMs);
      if ((lt - expLightS).abs() <= 2.0) {
        passed++;
      } else {
        failed++;
        print('FAIL: $tag lightTravel — expected ${expLightS.toStringAsFixed(3)}, got ${lt.toStringAsFixed(3)}');
      }
    }

    if (expPiw >= 0) {
      if (pt.periodInWeek == expPiw) {
        passed++;
      } else {
        failed++;
        print('FAIL: $tag period_in_week=$expPiw (got ${pt.periodInWeek})');
      }
    }

    if (expIsWorkPeriod >= 0) {
      final got = pt.isWorkPeriod ? 1 : 0;
      if (got == expIsWorkPeriod) {
        passed++;
      } else {
        failed++;
        print('FAIL: $tag is_work_period=$expIsWorkPeriod (got $got)');
      }
    }

    if (expIsWorkHour >= 0) {
      final got = pt.isWorkHour ? 1 : 0;
      if (got == expIsWorkHour) {
        passed++;
      } else {
        failed++;
        print('FAIL: $tag is_work_hour=$expIsWorkHour (got $got)');
      }
    }
  }

  print('Fixture entries checked: ${entries.length}');
  print('$passed passed  $failed failed');

  if (failed > 0) {
    exit(1);
  }
}

File: elixir/planet-time/fixture_runner/fixture_runner.exs
#!/usr/bin/env elixir
# fixture_runner.exs — validates 54 cross-language fixture entries
#
# Usage (via mix run — loads the compiled project, Jason available):
#   mix run fixture_runner/fixture_runner.exs ../../c/planet-time/fixtures/reference.json
#
# The script is invoked by `mix run` so all project modules are available.

path = System.argv() |> List.first() || "../../c/planet-time/fixtures/reference.json"

unless File.exists?(path) do
  IO.puts("SKIP: fixture file not found at #{path}")
  IO.puts("0 passed  0 failed  (fixtures skipped)")
  System.halt(0)
end

json_text = File.read!(path)

data =
  try do
    Jason.decode!(json_text)
  rescue
    e ->
      IO.puts(:stderr, "Failed to parse fixture: #{inspect(e)}")
      System.halt(1)
  end

entries = Map.fetch!(data, "entries")

# Ensure all planet atoms are registered in the atom table before
# String.to_existing_atom/1 is called for the first fixture entry.
_ = InterplanetTime.Constants.planets()

passed = 0
failed = 0

{passed, failed} =
  Enum.reduce(entries, {0, 0}, fn entry, {p, f} ->
    utc_ms     = trunc(entry["utc_ms"])
    planet_str = entry["planet"]
    exp_hour   = entry["hour"]
    exp_minute = entry["minute"]
    exp_lt     = entry["light_travel_s"]

    tag = "#{planet_str}@#{utc_ms}"

    planet =
      try do
        String.to_existing_atom(planet_str)
      rescue
        _ ->
          IO.puts("FAIL: #{tag} unknown planet: #{planet_str}")
          nil
      end

    if is_nil(planet) do
      {p, f + 1}
    else
      pt = InterplanetTime.get_planet_time(planet, utc_ms)

      {p1, f1} =
        if pt.hour == exp_hour do
          {p + 1, f}
        else
          IO.puts("FAIL: #{tag} hour=#{exp_hour} (got #{pt.hour})")
          {p, f + 1}
        end

      {p2, f2} =
        if pt.minute == exp_minute do
          {p1 + 1, f1}
        else
          IO.puts("FAIL: #{tag} minute=#{exp_minute} (got #{pt.minute})")
          {p1, f1 + 1}
        end

      {p3, f3} =
        if not is_nil(exp_lt) and exp_lt != 0.0 and
           planet_str not in ["earth", "moon"] do
          lt = InterplanetTime.light_travel_seconds(:earth, planet, utc_ms)
          if abs(lt - exp_lt) <= 2.0 do
            {p2 + 1, f2}
          else
            IO.puts("FAIL: #{tag} lightTravel — expected #{:erlang.float_to_binary(exp_lt, [{:decimals, 3}])}, got #{:erlang.float_to_binary(lt, [{:decimals, 3}])}")
            {p2, f2 + 1}
          end
        else
          {p2, f2}
        end

      {p3, f3}
    end
  end)

IO.puts("Fixture entries checked: #{length(entries)}")
IO.puts("#{passed} passed  #{failed} failed")

if failed > 0, do: System.halt(1)
